source/h2_frames.c

```diff
@@ -332,714 +256,829 @@ void aws_h2_frame_encoder_clean_up(struct aws_h2_frame_encoder *encoder) {
-    }
-    /* Write padding */
-    for (size_t i = 0; i < frame->pad_length; ++i) {
-        if (!aws_byte_buf_write_u8(output, 0)) {
-            goto write_error;
+    } else {
+        if (body_sub_buf.len == 0) {
+            /* This frame would have no useful information, don't even bother sending it */
+            goto handle_nothing_to_send_right_now;
         }
     }
 
-    return AWS_OP_SUCCESS;
-
-write_error:
-    output->len = output_init_len;
-    return aws_raise_error(AWS_H2_ERR_FRAME_SIZE_ERROR);
+    /*
+     * Write in the other parts of the frame.
+     */
+    bool all_wrote = true;
 
-compression_error:
-    output->len = output_init_len;
-    return aws_raise_error(AWS_ERROR_HTTP_COMPRESSION);
-}
+    /* Write the frame prefix */
+    const size_t payload_len = body_sub_buf.len + payload_overhead;
+    s_frame_prefix_encode(AWS_H2_FRAME_T_DATA, stream_id, payload_len, flags, output);
 
-/***********************************************************************************************************************
- * PRIORITY
- **********************************************************************************************************************/
-static const size_t s_frame_priority_length = 5;
+    /* Write pad length */
+    if (flags & AWS_H2_FRAME_F_PADDED) {
+        all_wrote &= aws_byte_buf_write_u8(output, pad_length);
+    }
 
-int aws_h2_frame_priority_init(struct aws_h2_frame_priority *frame, struct aws_allocator *allocator) {
-    (void)allocator;
+    /* Increment output->len to jump over the body that we already wrote in */
+    AWS_ASSERT(output->buffer + output->len == body_sub_buf.buffer && "Streamed DATA to wrong position");
+    output->len += body_sub_buf.len;
 
-    AWS_ZERO_STRUCT(*frame);
-    frame->base.type = AWS_H2_FRAME_T_PRIORITY;
+    /* Write padding */
+    if (flags & AWS_H2_FRAME_F_PADDED) {
+        all_wrote &= aws_byte_buf_write_u8_n(output, 0, pad_length);
+    }
 
+    AWS_ASSERT(all_wrote);
     return AWS_OP_SUCCESS;
-}
-void aws_h2_frame_priority_clean_up(struct aws_h2_frame_priority *frame) {
-    AWS_PRECONDITION(frame);
-
-    (void)frame;
-}
 
-int aws_h2_frame_priority_encode(
-    struct aws_h2_frame_priority *frame,
-    struct aws_h2_frame_encoder *encoder,
-    struct aws_byte_buf *output) {
-    AWS_PRECONDITION(frame);
-    AWS_PRECONDITION(encoder);
-    AWS_PRECONDITION(output);
-
-    (void)encoder;
-
-    const size_t output_init_len = output->len;
-
-    /* Write the header data */
-    if (s_frame_prefix_encode(&frame->base, s_frame_priority_length, 0, output)) {
-        goto write_error;
-    }
-
-    /* Write the priority settings */
-    if (s_frame_priority_settings_encode(&frame->priority, output)) {
-        goto write_error;
-    }
+handle_waiting_for_more_space:
+    ENCODER_LOGF(TRACE, encoder, "Insufficient space to encode DATA for stream %" PRIu32 " right now", stream_id);
+    return AWS_OP_SUCCESS;
 
+handle_nothing_to_send_right_now:
```

Comment:
did we ever figure out how we're going to handle this? Or are we just going to keep asking for data on ever event loop tick?

Reply:
Until we beef up the `aws_input_stream` API to be Reactive-Streams-esque, I plan to just tick every loop